pinot-common/src/main/java/org/apache/pinot/common/function/JsonPathMapCache.java

```diff
@@ -0,0 +1,61 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.common.function;
+
+import com.jayway.jsonpath.JsonPath;
+import com.jayway.jsonpath.spi.cache.Cache;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * This cache for avoiding repeatedly compiling JSON path string when using jsonPathXXX transformations
+ *
+ * <p>The existing jay way LRUCache is very inefficient, It will cause
+ * a lot of unnecessary lock waits during high concurrent data ingestion,
+ * and LRU mechanism is inappropriate for Pinot bounded size of the
+ * transformation config, so we should use this simple Map cache instead
+ * of it.
```

Comment:
JsonPath can also be used at query time (see `JsonExtractScalarTransformFunction`). Let's add a TODO here to add some evict policy in the future

Reply:
I think in the future this can be handled better by just precompiling `JsonPath` objects and bypassing the library's cache entirely. For transformConfig, it would require changes to the way `ScalarFunction` is defined as well as changes to `InbuiltFunctionEvaluator.planExecution`. For `JsonExtractScalarTransformFunction` the change is trivial. 

I think in the meantime using a guava `Cache` instead of a `ConcurrentHashMap` as suggested on slack would alleviate this concern.

Reply:
I think guava `Cache` is a more safe cache solution cause the `JsonPath` cache will also be applied to the query.
For one of my table ingestion cases, reading data QPS close to 10w and configuring `JsonPath` transformations with dozens of fields, this will put a certain pressure on GC and CPU because of the LRU algorithm needs to record every `get`. Of course, this should not become a system bottleneck, but the implementation is a bit too heavy.

Reply:
Guava Cache uses `recencyQueue` to record cache accesses. A large number of `CAS` enqueue operations generated by hotkeys will cause CPU consumption, and this synchronous recording will also slow down cache reading.

I think maybe [Caffeine Cache](https://github.com/ben-manes/caffeine) is a good choice. Its Striped-RingBuffer design is more efficient for GC and CPU usage. And its API is very similar to Guava Cache.

Reply:
Whether using Guava's cache would result in high CPU consumption needs measurement. I agree that Caffeine is a better cache implementation, but Guava is already on the classpath. Is this particular use case important enough to add a 1MB depedency?

Reply:
You will likely be fine using Guava if you increase the concurrency level (e.g. to 64, defaults to 4). That will stripe the LRU so that there is less contention, but will still suffer at peak loads ([benchmarks](https://github.com/ben-manes/caffeine/wiki/Benchmarks#read-100-1)). For your use-case here that is likely good enough given other costs, e.g. JsonPath's [inefficiencies](https://github.com/json-path/JsonPath/issues/740). Unfortunately my patches to Guava that gave it a ~25x boost were not accepted due to a lack of a champion, and instead they simply updated the JavaDoc to recommend Caffeine. 🙂

Reply:
I think we should use the default value of the `concurrencyLevel` of `Guava cache`. In this scenario, there will be very few caches writes and no need to update the cache. My concern is the way Guava records cache reads, which is not very efficient.

Reply:
The performance of cache reads improves with the concurrency level due to reducing the likelihood of contending on the same CLQ instance. The benchmarks use a zipf to simulate a hotspot.

Reply:
You are right.

Reply:
Thanks for the insights @ben-manes. FWIW I have been using Caffeine in commercial projects for years, the only question here is whether _this_ particular use case is worth adding a relatively large (in bytes) dependency.

Reply:
It's not a problem. I co-authored Guava's so either way my code is used. 😄 

@Ferrari6 is correct in what your preference and default should be, though. Prior to my involvement, the Guava team bet on reference caching (MapMaker) and spent their complexity budget by forking the hash table as an optimization. This was a mistake because soft references can cause GC death spirals of stop the world events and unpredictable evictions, but looked fine in a naive benchmark. This blew their complexity budget, so porting size eviction from [CLHM](https://github.com/ben-manes/concurrentlinkedhashmap) favored simplicity over performance.

The longer-term problem you'll face is that no one maintains CacheBuilder. The last big change was adding Map.compute, but it was riddled with major bugs and done inefficiently. I fixed some of those problems, but there are show stoppers. If you keep to the historic functionality then Guava's can be made acceptable in most cases. Caffeine does have jar bloat by code generating per-configuration entry classes to minimize the memory footprint. In those cases where disk is a premium some projects embed CLHM into their code base, e.g. msjdbc and groovy. You'll probably have many caches throughout Pinot making the Caffeine dependency worthwhile even if you can get by in a case-by-case basis.